### DISCUSSION TITLE
Thread safety of ISerializer/IDeserializer

### DISCUSSION THREAD

**Discussion by filzrev** [aaubry/YamlDotNet#844] [Q&A] [answered]
I want to know ISerializer/IDeserializer is thread-safe or not thread-safe.

On [this StackOverflow question](https://stackoverflow.com/questions/70696315/can-yamldotnets-ideserializer-be-used-as-a-singleton-in-iservicecollection/71241271#71241271).  
Project author answered it's safe to be called by multiple threads.

But it can be found some comments that it's not thread safe.
- https://github.com/aaubry/YamlDotNet/discussions/799
- https://github.com/aaubry/YamlDotNet/issues/517#issuecomment-1620451059

Is there any official documents about multi thread behavior?

**Comment by EdwardCooke** [comment, marked as answer]
No, they are not thread safe. You will need to create new serializers and deserializers in each thread.

**Reply by filzrev** [reply]
Thanks for your clarification.

Since serializer/deserializer API is not thread-safe. (SerializerBuilder/DeserializerBuilder also?)
I'll try to create serializer/deserializer per-thread and cache instance by using `ThreadLocal` functionality.

```cs
    private static readonly ThreadLocal<ISerializer> Serializer = new(() =>
      new SerializerBuilder()
      .WithNamingConvention(CamelCaseNamingConvention.Instance)
      .Build()
    );
```

**Comment by EdwardCooke** [comment]
It's thread safe now as of 16.1.0
